### PR TITLE
Feature/fixed grid pchip fourier integral

### PR DESCRIPTION
**Pull request by veskil**
Depends on #1

Copy of #2 to get comparison against current state of main.

Implements a function `fourier_integral_fixed_sampling_pchip` which takes an array of frequencies $f$ and an array of corresponding function outputs for some function $g(f)$ and calculates the fourier integral of $g$ at times $t$ also given as an input.

The calculation approximates the function as a piecewise cubic Hermite interpolating polynomial (PCHIP) and calculates the fourier integral analytically for this interpolating polynomial.

Heavily based on section 1.6.3 in [N. Mounet. The LHC Transverse Coupled-Bunch Instability, PhD thesis 5305 (EPFL, 2012)](https://wiki.epfl.ch/nmounet/documents/phd_thesis_nmounet.pdf)

### CODE REVIEW

File: neffint/fixed_grid_fourier_integral.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    :param times: A 1D array of length M with the times to compute the Fourier integral for
+    :type times: np.ndarray
+    :param omega_end: A frequency to expand the Taylor series from
+    :type omega_end: float
+    :param func_value_end: An array of shape (X1, X2, ...) containing the values of the function to be transformed evaluated at omega_end
+    :type func_value_end: ArrayLike
+    :param func_derivative_end: An array of shape (X1, X2, ...) containing the derivative of the function to be transformed, evaluated at omega_end, defaults to 0.
+    :type func_derivative_end: ArrayLike, optional
+    :return: The asymptotic terms of the Fourier Integral for all times, given as an array of shape (M, X1, X2, ...)
+    :rtype: np.ndarray
```
Comment (veskil) [inline neffint/fixed_grid_fourier_integral.py:148]: I did not mention units in the docstring (yet). I think it is probably a good idea to mention it, but not sure if I should just write [s] and [Hz] or if I should write that any coherent time and frequency units work (e.g. ns and GHz should work, so should m and m^-1)

File: tests/test_fixed_grid_fourier_integral.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        -0.0833333333333333333 - 5e-11j,
+        -0.06739546857228461362 - 0.046145700566923663667j,
+        8.3788717918052853757e-19 + 5.4584345480024828642e-19j,
+        1.1837199021871073658e-37 - 9.9296932074040507374e-37j
+    ])
+
+    output_phi, output_psi = _phi_and_psi(input_x)
+
+    assert output_phi == pytest.approx(expected_phi, rel=comparison_tolerance, abs=comparison_tolerance)
+    assert output_psi == pytest.approx(expected_psi, rel=comparison_tolerance, abs=comparison_tolerance)
```
Comment (veskil) [inline tests/test_fixed_grid_fourier_integral.py:50]: Let me know if I should test on more inputs. My thought of having only 4 here is that it is fairly short and simple, but could easily do more. Main downside would be that the expected outputs take up more space on the screen, making it harder to read.

File: tests/test_fixed_grid_fourier_integral.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    output_transform_arr = fourier_integral_fixed_sampling_pchip(
+        times=input_times,
+        frequencies=input_frequencies,
+        func_values=input_func_arr,
+        inf_correction_term=True,
+    )
+
+    expected_transform_arr = np.array([expected_transform(time) for time in input_times])
+
+    assert np.real(output_transform_arr) == pytest.approx(expected_transform_arr, rel=1e-4)
```
Comment (veskil) [inline tests/test_fixed_grid_fourier_integral.py:74]: Have to accept quite high relative error here with this number of frequencies for this frequency range and time range. See https://github.com/neffint/neffint/pull/2#issuecomment-1520132201

**Review by veskil — COMMENTED** [review]
Bring in code comments from #2

**Comment by veskil** [comment]
## Performance

Using realistic intermediate value from the C++ code, i.e. by taking a snapshot of all arrays after the bisection algorithm has finished, this is how the newly implemented functions perform compared to IW2D. This is using `IW2D/examples/input_files/WakeFlatChamberInputFile.txt` with the tolerance changed to 1e13

### Phi and psi

image


### Fixed grid fourier integral

Disclaimer: I'm a little worried this relatively large error is partly due to how I extract the numbers from C++, will look more into it

image

---

EDIT: So as far as I can tell, the reason for the error is that scipy's pchip and the IW2D pchip apparently disagree quite a bit, with a similar error magnitude as in transform above. Will investigate further

---

EDIT2: So the discrepancy in PCHIP derivatives seems indeed to be the primary reason for the large differences between the two. I will put up an issue detailing it later. Added a commit where I changed the PCHIP to treat the real and complex parts separately. This is not what IW2D does, so the results are still quite different, but now I think that is due to the way it is implemented in IW2D, and not any bugs in the implementation here.

---

EDIT 3: So as to not only compare to IW2D - here are some plots comparing the results against analytically computed integral, more specifically $\frac{1}{\sqrt{\omega}}$:

1k frequencies
invsqrt_pchip_1000f

10k frequencies
invsqrt_pchip_10kf

New input and output range, 1k frequencies
invsqrt_pchip_1000f_newrange

10k frequencies
invsqrt_pchip_10kf_newrange

File: tests/test_fixed_grid_fourier_integral.py (hunk 2 above)
Comment (nmounet) [reply tests/test_fixed_grid_fourier_integral.py:50]: it is good like that!

File: tests/test_fixed_grid_fourier_integral.py (hunk 3 above)
Comment (nmounet) [reply tests/test_fixed_grid_fourier_integral.py:74]: I think it's ok

File: tests/test_fixed_grid_fourier_integral.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    assert output_phi == pytest.approx(expected_phi, rel=comparison_tolerance, abs=comparison_tolerance)
+    assert output_psi == pytest.approx(expected_psi, rel=comparison_tolerance, abs=comparison_tolerance)
+
+
+@pytest.mark.parametrize(("input_func", "expected_transform"), [
+    ( lambda f: 1 / np.sqrt( 2 * np.pi * f),                        lambda t: np.sqrt(np.pi / (2 * t)) ),
+    ( lambda f: np.array([[1,2],[3,4]]) / np.sqrt( 2 * np.pi * f),  lambda t: np.array([[1,2],[3,4]]) * np.sqrt(np.pi / (2 * t)) ), # Test dimension handling
+    ( lambda f: 1 / (1 + (2 * np.pi * f)**2),                       lambda t: np.pi / 2 * np.exp(-t) ),
+])
+def test_fourier_integral_fixed_sampling_pchip(input_func: Callable[[ArrayLike], ArrayLike], expected_transform: Callable[[ArrayLike], ArrayLike]):
```
Comment (nmounet) [inline tests/test_fixed_grid_fourier_integral.py:58]: if we implement the extrapolation to negative frequencies, one test should be added specifically for this case (or we just test separately the asymptotic formula)

**Review by nmounet — COMMENTED** [review]
wonderful work. I like a lot the very pythonic way in which additional dimensions are handled.

comments are mainly about some minor typos, or docstring details. There is also a comment about adding the -infinity part of the integral - can be considered now or in a later PR.

File: neffint/fixed_grid_fourier_integral.py (hunk 1 above)
Comment (veskil) [reply neffint/fixed_grid_fourier_integral.py:148]: I agree including the units you suggest. Might annotate them with a `*` and add a footnote like

    * Any coherent time and frequency units will work.

PS: I see I copied the review comment from the wrong docstring, hence the `2pi` mixup

File: tests/test_fixed_grid_fourier_integral.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+import numpy as np
+import pytest
+from numpy.typing import ArrayLike
+
+from neffint.fixed_grid_fourier_integral import (
+    fourier_integral_fixed_sampling_pchip, _phi_and_psi)
+
+
+def test_phi_and_psi():
```
[the remark this answers is not in the document]
Comment (veskil) [reply tests/test_fixed_grid_fourier_integral.py:25]: The reasoning for not parametrizing is that now we test that it does the "split" correctly. I.e. that when given some `x > 1` and some `x < 1`, it uses the Taylor series for the small inputs. The input will almost always be an array in real situations anyway.

File: neffint/fixed_grid_fourier_integral.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    
+    [1] F. N. Fritsch and J. Butland, A method for constructing local monotone piecewise cubic interpolants, SIAM J. Sci. Comput., 5(2), 300-304 (1984). DOI:10.1137/0905021.
+
+    :param xi: A 1D array of N real input points
+    :type xi: ArrayLike
+    :param func_values: An array of of shape (N, X1, X2, ...) containing the output of the function to compute the interpolation for at each xi.
+    :type func_values: ArrayLike
+    :param x: A 1D array of M real inputs to evaluate the interpolation at
+    :type x: ArrayLike
+    :param derivative_order: The order of derivatives to compute, defaults to 0
```
[the remark this answers is not in the document]
Comment (veskil) [reply neffint/fixed_grid_fourier_integral.py:41]: I agree with both

File: tests/test_fixed_grid_fourier_integral.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nmounet) [reply tests/test_fixed_grid_fourier_integral.py:75]: you are completely right, I overlooked this, indeed :thumbsup:

File: neffint/fixed_grid_fourier_integral.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+        abs_x_to_n_plus_1 *= np.abs(xx)
+        inv_factorial /= n+1
+
+    # If loop finished without converging, raise error
+    if not (phi_converged and psi_converged):
+        raise RuntimeError(f"Phi and Psi calculation failed to converge after {MAX_PHI_PSI_ITERATIONS} iterations")
+    
+    return phi, psi
+
+def fourier_integral_inf_correction(times: ArrayLike, omega_end: float, func_value_end: ArrayLike, func_derivative_end: ArrayLike=0., positive_inf: bool=True) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (veskil) [reply neffint/fixed_grid_fourier_integral.py]: Indeed! Fixed now

**Pull request merged**